kafka-connect/src/main/java/io/tabular/iceberg/connect/channel/Deduplicated.java

```diff
@@ -0,0 +1,236 @@
+   * about the ordering of the files that are returned.
+   */
+  public static List<DataFile> dataFiles(
+      UUID currentCommitId, TableIdentifier tableIdentifier, List<Envelope> envelopes) {
+    return deduplicatedFiles(
+        currentCommitId,
+        tableIdentifier,
+        envelopes,
+        "data",
+        CommitResponsePayload::dataFiles,
+        dataFile -> dataFile.path().toString());
+  }
+
+  /**
+   * Returns the deduplicated delete files from the batch of envelopes. Does not guarantee anything
+   * about the ordering of the files that are returned.
+   */
+  public static List<DeleteFile> deleteFiles(
+      UUID currentCommitId, TableIdentifier tableIdentifier, List<Envelope> envelopes) {
+    return deduplicatedFiles(
+        currentCommitId,
+        tableIdentifier,
+        envelopes,
+        "delete",
+        CommitResponsePayload::deleteFiles,
+        deleteFile -> deleteFile.path().toString());
+  }
+
+  private static <F> List<F> deduplicatedFiles(
+      UUID currentCommitId,
+      TableIdentifier tableIdentifier,
+      List<Envelope> envelopes,
+      String fileType,
+      Function<CommitResponsePayload, List<F>> extractFilesFromPayload,
+      Function<F, String> extractPathFromFile) {
+    List<FileAndEnvelope<F>> filesAndEnvelopes =
+        envelopes.stream()
+            .flatMap(
+                envelope -> {
+                  CommitResponsePayload payload =
+                      (CommitResponsePayload) envelope.event().payload();
+                  List<F> files = extractFilesFromPayload.apply(payload);
+                  if (files == null) {
+                    return Stream.empty();
+                  } else {
+                    return deduplicate(
+                            files,
+                            extractPathFromFile,
+                            (path, duplicateFiles) ->
+                                detectedDuplicateFilesInSameEventMessage(
+                                    path,
+                                    duplicateFiles,
+                                    fileType,
+                                    envelope,
+                                    tableIdentifier,
+                                    currentCommitId))
+                        .stream()
+                        .map(file -> new FileAndEnvelope<>(file, envelope));
+                  }
+                })
+            .collect(toList());
+
+    List<FileAndEnvelope<F>> result =
+        deduplicate(
+            filesAndEnvelopes,
+            fileAndEnvelope -> extractPathFromFile.apply(fileAndEnvelope.getFile()),
+            (path, duplicateFilesAndEnvelopes) ->
+                detectedDuplicateFilesAcrossMultipleEventsMessage(
+                    path, duplicateFilesAndEnvelopes, fileType, tableIdentifier, currentCommitId));
+
+    return result.stream().map(FileAndEnvelope::getFile).collect(toList());
+  }
+
+  private static class FileAndEnvelope<T> {
```

Comment:
Nit.  Nested classes at the bottom of the file. 


This would be easier if we had record types (or tuples) but unfortunately we are trapped in JDK8.

Might be easier to use `org.apache.iceberg.util.Pair` which is in scope and remove this nested class entirely.

Reply:
Good idea, didn't see it in scope